### PR TITLE
Add basic fabric link monitoring counters and states handling.

### PR DESCRIPTION
**Pull request by jfeng-arista**
**What I did**
Added basic fabric link monitoring counters and states handling, and the test for this. 
**Why I did it**

**How I verified it**

**Details if related**

### CODE REVIEW

**Comment by jfeng-arista** [comment]
it looks like none of the test_fabric* run

**Comment by jfeng-arista** [comment]
I checked log@1 from https://dev.azure.com/mssonic/build/_build/results?buildId=431467&view=artifacts&pathAsName=false&type=publishedArtifacts

from the syslog of all test_fabric_* , I can see FabricPortsOrch get construct and at least generatePortStats getting called and etc.  so at least it seems the orchagent.fabricportsorch.cpp get covered, should not get 0% coverage here. 

Would anyone please  help check this ? and pointing what actually missed ? thanks

**Comment by jfeng-arista** [comment]
test_add_remove_port failed (1 runs remaining out of 2).
	<class 'AssertionError'>
	assert False == True
  -False
  +True
	[<TracebackEntry /agent/_work/1/s/tests/test_drop_counters.py:726>]
test_add_remove_port failed; it passed 0 out of the required 1 times.
	<class 'AssertionError'>
	assert 1 == 0
  -1
  +0
	[<TracebackEntry /agent/_work/1/s/tests/test_drop_counters.py:688>]
test_createAndDeleteMultipleCounters failed (1 runs remaining out of 2).
	<class 'AssertionError'>
	assert 'SAI_PORT_STAT_IN_CONFIGURED_DROP_REASONS_1_DROPPED_PKTS' in 'SAI_PORT_STAT_OUT_CONFIGURED_DROP_REASONS_1_DROPPED_PKTS,SAI_PORT_STAT_IN_CONFIGURED_DROP_REASONS_0_DROPPED_PKTS'
	[<TracebackEntry /agent/_work/1/s/tests/test_drop_counters.py:762>, <TracebackEntry /agent/_work/1/s/tests/test_drop_counters.py:109>]
test_createAndDeleteMultipleCounters failed; it passed 0 out of the required 1 times.
	<class 'AssertionError'>
	assert 'SAI_PORT_STAT_IN_CONFIGURED_DROP_REASONS_1_DROPPED_PKTS' in 'SAI_PORT_STAT_OUT_CONFIGURED_DROP_REASONS_1_DROPPED_PKTS,SAI_PORT_STAT_IN_CONFIGURED_DROP_REASONS_0_DROPPED_PKTS'
	[<TracebackEntry /agent/_work/1/s/tests/test_drop_counters.py:762>, <TracebackEntry /agent/_work/1/s/tests/test_drop_counters.py:109>]

It is not caused by this change .

**Comment by jfeng-arista** [comment]
three bfd related tests failed

**Comment by jfeng-arista** [comment]
thank you

**Comment by jfeng-arista** [comment]
test_NeighborAddRemoveIpv6LinkLocal failed, does not related to this change

**Comment by xumia** [comment]
For code coverage, please skip the check coverage.Azure.sonic-swss.amd64, only need to pass coverage.Azure.sonic-swss.vstest.

**Comment by kenneth-arista** [comment]
@arlakshm

**Comment by prsunny** [comment]
@arlakshm , @abdosi , please signoff

**Comment by judyjoseph** [comment]
@vmittal-msft @saksarav-nokia please review

**Review by arlakshm — COMMENTED** [review]
lgtm, few minor comments

File: orchagent/fabricportsorch.cpp [hunk 1]
```diff
@@ -348,13 +742,28 @@ void FabricPortsOrch::doTask(swss::SelectableTimer &timer)
...
+            updateFabricPortState();
+        }
+    }
+    else if (timer.getFd() == m_debugTimer->getFd())
     {
-        updateFabricPortState();
+        if (!m_getFabricPortListDone)
+        {
+            getFabricPortList();
+        }
```
[the remark this answers is not in the document]
Comment (jfeng-arista) [reply orchagent/fabricportsorch.cpp:762]: > why we need to get the fabric port list again?

we try to catch up the case if getFabricPortList did not get called before or not successfully get the done so m_getFabricPortListDone get set. Here we only try to call getFabricPortList again if  m_getFabricPortListDone is false.  thank you

File: orchagent/fabricportsorch.cpp [hunk 2]
```diff
@@ -336,6 +358,378 @@ void FabricPortsOrch::updateFabricPortState()
...
+        string lnkStatus = "down";
+        string testState = "product";
+
+        // Get the consecutive polls from the state db
+        std::vector<FieldValueTuple> values;
+        string valuePt;
+        bool exist = m_stateTable->get(key, values);
+        if(!exist)
+        {
+           SWSS_LOG_INFO("No state infor for port %s", key.c_str());
```
[the remark this answers is not in the document]
Comment (jfeng-arista) [reply orchagent/fabricportsorch.cpp:477]: > fix identation

fixed. thank you!

File: orchagent/fabricportsorch.cpp [hunk 3]
```diff
@@ -336,6 +358,378 @@ void FabricPortsOrch::updateFabricPortState()
...
+           if (fvField(val) == "TEST_CODE_ERRORS")
+           {
+              testCodeErrors = to_uint<uint64_t>(valuePt);
+              continue;
+           }
+           if (fvField(val) == "TEST")
+           {
+              testState = valuePt;
+              continue;
+           }
```
[the remark this answers is not in the document]
Comment (jfeng-arista) [reply orchagent/fabricportsorch.cpp:553]: > Fix identation

fixed. thank you !

File: orchagent/fabricportsorch.cpp [hunk 4]
```diff
@@ -336,6 +358,378 @@ void FabricPortsOrch::updateFabricPortState()
...
+           if (autoIsolated == 0 && (consecutivePollsWithErrors >= isolationPollsCfg
+                                  || consecutivePollsWithFecErrs >= fecIsolatedPolls))
+           {
+              // Link needs to be isolated.
+              SWSS_LOG_INFO("port %s auto isolated", key.c_str());
+              autoIsolated = 1;
+              valuePt = to_string(autoIsolated);
+              m_stateTable->hset(key, "AUTO_ISOLATED", valuePt);
+              SWSS_LOG_INFO("port %s set AUTO_ISOLATED %s", key.c_str(), valuePt.c_str());
+              // Call SAI api here to actually isolated the link
```
[the remark this answers is not in the document]
Comment (jfeng-arista) [reply orchagent/fabricportsorch.cpp:679]: > This is will be subsequent PR?

Yes. There will be subsequent PR of this actual sai call when it is available   .  thank you

File: orchagent/fabricportsorch.cpp [hunk 5]
```diff
@@ -336,6 +358,378 @@ void FabricPortsOrch::updateFabricPortState()
...
+           SWSS_LOG_INFO("consecutivePollsWithFecErrs %d consecutivePollsWithNoFecErrs %d",
+                         consecutivePollsWithFecErrs,consecutivePollsWithNoFecErrs);
+           SWSS_LOG_INFO("fecUnisolatePolls %d", fecUnisolatePolls);
+        }
+
+        // take care serdes link shut state setting
+        if (lnkStatus == "up")
+        {
+           // debug information
+           SWSS_LOG_INFO("port %s status up autoIsolated %d",
```
[the remark this answers is not in the document]
Comment (jfeng-arista) [reply orchagent/fabricportsorch.cpp:664]: > can we make this log as NOTICE

Here is the debug information for all links and that might be too chatty. BUT I did change the log level to NOTICE several lines down where we decide to isolate the link due to the problem or unisolate due to the link healthy again, so we will have the information for links that in trouble. 

           {
              // Link needs to be isolated.
              SWSS_LOG_INFO("port %s auto isolated", key.c_str());
              autoIsolated = 1;
              valuePt = to_string(autoIsolated);
              m_stateTable->hset(key, "AUTO_ISOLATED", valuePt);
              **SWSS_LOG_NOTICE**("port %s set AUTO_ISOLATED %s", key.c_str(), valuePt.c_str());
              // Call SAI api here to actually isolated the link
           }

           {
              // Link is isolated, but no longer needs to be.
              SWSS_LOG_INFO("port %s healthy again", key.c_str());
              autoIsolated = 0;
              valuePt = to_string(autoIsolated);
              m_stateTable->hset(key, "AUTO_ISOLATED", valuePt);
              **SWSS_LOG_NOTICE**("port %s set AUTO_ISOLATED %s", key.c_str(), valuePt.c_str());
              // Can we call SAI api here to unisolate the link?
           }

File: orchagent/fabricportsorch.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (jfeng-arista) [reply orchagent/fabricportsorch.cpp:761]: > why we need to get the fabric port list again?

Discussed with Arvind, remove this getFabricPortList call here. thank you

**Review by arlakshm — APPROVED** [review]

**Pull request merged**